File: java/sorting/standard/QuickSort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class QuickSort {

    public static int[] quickSort(int[] arr) {
        if (arr.length <= 1) {
            return Arrays.copyOf(arr, arr.length);
        }

        int pivot = arr[arr.length / 2];
        List<Integer> left = new ArrayList<>();
        List<Integer> middle = new ArrayList<>();
        List<Integer> right = new ArrayList<>();

        for (int value : arr) {
            if (value < pivot) {
                left.add(value);
            } else if (value > pivot) {
                right.add(value);
            } else {
                middle.add(value);
            }
        }

        return concatenate(
                quickSort(left.stream().mapToInt(i -> i).toArray()),
                middle.stream().mapToInt(i -> i).toArray(),
                quickSort(right.stream().mapToInt(i -> i).toArray())
        );
    }

    private static int[] concatenate(int[] left, int[] middle, int[] right) {
        int[] result = new int[left.length + middle.length + right.length];
        int index = 0;

        for (int value : left) {
            result[index++] = value;
        }
        for (int value : middle) {
            result[index++] = value;
        }
        for (int value : right) {
            result[index++] = value;
        }

        return result;
    }
}

File: cpp/sorting/standard/quick_sort.cpp
#include <vector>

using namespace std;

vector<int> quick_sort(const vector<int>& arr) {
    if (arr.size() <= 1) return arr;

    int pivot = arr[arr.size() / 2];
    vector<int> left, middle, right;
    for (int x : arr) {
        if (x < pivot) left.push_back(x);
        else if (x == pivot) middle.push_back(x);
        else right.push_back(x);
    }

    vector<int> result;
    vector<int> sorted_left = quick_sort(left);
    vector<int> sorted_right = quick_sort(right);
    result.insert(result.end(), sorted_left.begin(), sorted_left.end());
    result.insert(result.end(), middle.begin(), middle.end());
    result.insert(result.end(), sorted_right.begin(), sorted_right.end());
    return result;
}
